Applications/WebPortal/WebApp/library/source/pages/account_directory_page/cached_account_directory_model.tsx

```diff
@@ -0,0 +1,65 @@
+import * as Beam from 'beam';
+import { AccountDirectoryModel } from './account_directory_model';
+import { AccountEntry } from './account_entry';
+
+/** Implements an AccountDirectoryModel that caches. */
+export class CachedAccountDirectoryModel extends AccountDirectoryModel {
+
+  constructor(model: AccountDirectoryModel) {
+    super();
+    this._model = model;
+    this._accounts = new Beam.Map<Beam.DirectoryEntry, AccountEntry[]>();
+    this._prevFiltered = new
+      Map<string,Beam.Map<Beam.DirectoryEntry, AccountEntry[]>>();
+  }
+
+  public get groups(): Beam.Set<Beam.DirectoryEntry> {
+    return this._model.groups;
+  }
+
+  public async loadAccounts(
+      group: Beam.DirectoryEntry): Promise<AccountEntry[]> {
+    if(!this._accounts.get(group)) {
+      this._accounts.set(group, await this._model.loadAccounts(group));
+    }
+    return this._accounts.get(group);
+  }
+
+  public async loadFilteredAccounts(filter: string):
+      Promise<Beam.Map<Beam.DirectoryEntry, AccountEntry[]>> {
+    let substringLength = filter.length;
+    while(substringLength) {
+      const substring = filter.substring(0, substringLength);
+      if(this._prevFiltered.get(substring)) {
+        if(filter !== substring) {
+          const superset = this._prevFiltered.get(substring);
```

Comment:
Don't duplicate the call to ```get```, you already called ```get``` on line 33, you will need to reuse the result of that call.